include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+{
+
+NBL_CONSTEXPR uint32_t BucketsPerThread = ceil((float) BucketCount / WorkgroupSize);
+
+groupshared uint32_t prefixScratch[BucketCount];
+
+struct ScratchProxy
+{
+    uint32_t get(const uint32_t ix)
+    {
+        return prefixScratch[ix];
+    }
+    void set(const uint32_t ix, const uint32_t value)
+    {
+        prefixScratch[ix] = value;
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
+
+template<typename KeyAccessor, typename ValueAccessor, typename ScratchAccessor>
+struct counting
+{
+    void histogram(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            int j = index + i * WorkgroupSize + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            nbl::hlsl::glsl::atomicAdd(sdata[k - data.minimum], (uint32_t) 1);
+        }
+
+        nbl::hlsl::glsl::barrier();
+
+        uint32_t sum = 0;
+        uint32_t scan_sum = 0;
+
+        for (int i = 0; i < BucketsPerThread; i++)
+        {
+            sum = nbl::hlsl::workgroup::exclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
+            template __call <ScratchProxy>
+            (sdata[WorkgroupSize * i + tid], arithmeticAccessor);
+
+            arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+            nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i + tid), sum);
+            if ((tid == WorkgroupSize - 1) && i > 0)
+                nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i), scan_sum);
+
+            arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+            if ((tid == WorkgroupSize - 1) && i < (BucketsPerThread - 1))
+            {
+                scan_sum = sum + sdata[WorkgroupSize * i + tid];
+                sdata[WorkgroupSize * (i + 1)] += scan_sum;
+            }
+        }
+    }
+                
+    void scatter(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ValueAccessor) val, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
```

Comment:
you need to use a `scratch` accessor here instead of the globally declared scratch array.

Furthemore you forgot about the optimization where we reduce the number of `atomic` calls on the global `histogram` by re-histogrammming the input keys, and then "peeling off" (subtracting) an appropriate number of elements from the global `histogram`

Reply:
Done